design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,188 @@
+# Overview
+This proposal introduces a mechanism to persist the firmware UUID of a Virtual Machine Instance (VMI) in KubeVirt.
+By storing the firmware UUID, we ensure that it remains consistent across VMI restarts.
+which is crucial for applications and services that rely on the UUID for identification or licensing purposes.
+
+## Motivation
+* Duplicate UUIDs cause burden with third-party automation frameworks such as Gitops tools, making it difficult to maintain unique VM UUID. 
+  Such tools have to manually assign unique UUIDs to VMs, thus also manage them externally. (issue)
+* The non-persistent nature of UUIDs may trigger redundant cloud-init setup upon VM restore. (issue)
+* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.
+
+## Goals
+* Improve the uniqueness of the VMI firmware UUID to become independent of the VM name and namespace.
+* Maintain the UUID persistence over the VM lifecycle.
+* Maintain backward compatibility
+* Maintain UUID persistence with VM backup/restore.
+
+## Non Goals
+
+
+## Definition of Users
+End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.
+
+## User Stories
+As an end-user, I expect my VMI to maintain its identity across restarts.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+
+This section provides a range of alternative solutions for implementing a persistent,
+universally unique firmware UUID for Virtual Machines Instances (VMIs) in KubeVirt.
+
+### 1. Persist UUID in the VM's Spec Field
+
+**Description:** The UUID is generated when the VMI is being created and is then saved to the VM's spec field.
+
+**Pros:**
+- Straightforward implementation.
+- Avoids the need to introduce new API fields.
+- Fully backward compatible.
+
+**Cons:**
+- Potential misuse of the spec field, which ideally should only reflect the user’s intended configuration.
+- The majority of users may not consider the UUID part of the desired VM state, making this field irrelevant from their perspective.
+- Can lead to longer and more complex VM definitions, impacting readability and management.
+
+---
+
+### 2. Generate a Default UUID via Webhook and Set it to the Spec Field
+
+**Description:** A webhook generates a UUID if the user does not provide one. This UUID is then set in the VM's spec field.
+
+**Pros:**
+- Retains backward compatibility and avoids introducing new API fields.
+
+**Cons:**
+- Same concerns as above about the spec field.
+- Using webhooks to assign defaults can degrade performance, especially with high volumes of VMs, as `virt-api` may become a bottleneck.
+- May compromise scalability, as generating a UUID via webhook adds overhead when multiple VMs start simultaneously.
+
+---
```

Comment:
You should describe why is it backward compatible (The controller uses same function to "generate" UUID in case of old vm where old vm is equal to vm that does not have the UUID set)